torch/testing/_internal/common_utils.py

```diff
@@ -691,12 +689,121 @@ def read_and_process():
             "Test is disabled because an issue exists disabling it: {}".format(disabled_test_from_issues[test_name]) +
             " To enable set the environment variable PYTORCH_RUN_DISABLED_TESTS=1")
 
+# Helper that returns the maximum value a dtype can represent
+def _get_width(dtype):
+    if dtype is torch.complex128:
+        return torch.finfo(torch.float64).bits
+    elif dtype is torch.complex64:
+        return torch.finfo(torch.float32).bits
+    elif dtype is torch.bool:
+        return 1
+    elif is_integral(dtype):
+        return torch.iinfo(dtype).bits
+    return torch.finfo(dtype).bits
+
+# Returns a dtype suitable for comparing tensors a and b.
+# NOTE: Dtypes for Comparing Tensors in the Test Framework
+# When two tensors are compared they may be of different dtypes.
+# In PyTorch, the typical type promotion rules are biased towards float
+# types, so tensors with dtypes float16 and int64 produce tensors of
+# dtype float16 when added together.
+# If we used typical PyTorch type promotion to determine the dtype that tensors
+# should be compared in, however, then the test framework would assert
+# that tensors which are actually very different were "equal" to each other.
+# For example, torch.tensor((50000), dtype=torch.long) would be "equal" to
+# torch.tensor(49984., dtype=torch.float16), regardless of the rtol and atol.
+# These cases are admittedly, unlikely, but the test framework is not intended
+# to reflect only "likely" scenarios.
+# To preserve numerical accuracy, instead of following PyTorch's type
+# promotion rules this function follows NumPy's, which are intended to preserve
```

Comment:
I think this is still a confusing explanation of the below code, in part because there are other differences that could be described as "preserves numerics." E.g. in numpy, a dtype is often found to preserve the actual inspected value regardless of dtypes:
```
>>> np.array([1], dtype=np.uint16) + np.array(1, dtype=np.int16)
array([2], dtype=uint16)
>>> np.array([1], dtype=np.uint16) + np.array(-1, dtype=np.int16)
array([0], dtype=int32)
```
If I check the implementation of promote_types in numpy, it doesn't seem like this is a direct port. It seems complicated enough that, if the goal is really to match np.promote_types, we'd need a test to ensure it does.

I really don't love adding a new slightly different variant of type promotion logic to maintain (and I suspect it's unlikely that we'll add a required numpy dependency), but another test could also capture if a new dtype is added and unhandled here.

Reply:
I did add a test called test_get_comparison_dtype in test_torch.py, but you're correct that the handling of scalars / wrapped scalars is not the same as NumPy (the function treats all scalars / wrapped scalars as tensors). I'll update the comment to reflect that.

I understand this function is a pain point, but I don't want to take a NumPy dependency in this PR. In the future we'll take one and I'll replace this with np.promote_types().

Reply:
Per later comment: I will replace this function with a table.

Reply:
sorry, I hadn't checked the tests yet.

I'd be surprised if we do eventually take a numpy dependency (I'm not sure if you're aware of a plan to?), so I'm assuming this will need to be maintained indefinitely.

I can see the value in having a way to get the kind of more precise comparison dtype numpy gives us (for a number of purposes). Maybe we'd want to expose promotion rules like this to users to allow control in mixed-precision ops, maybe as a flag on torch.promote_types. I'm just worried about maintainability/readability. There's a lot here and I wouldn't want to block the PR, but I'm also thinking a lot of this could end up getting re-written as we add more type metadata, e.g. for a subdtype api or other casting rules, and in the meantime could be a pain-point.

Reply:
True, but I also can't think of a good way around it atm :(

Reply:
> you're correct that the handling of scalars / wrapped scalars is not the same as NumPy (the function treats all scalars / wrapped scalars as tensors). I'll update the comment to reflect that

Also, the handling of scalars isn't really what I meant to clarify, my point there was just that if someone checks the numpy docs for what you mean by "numerical accuracy" they might get the wrong idea. I think we should give some concise conceptual explanation of what dtype we're returning. e.g. maybe it's something like "return a dtype that can exactly hold any value representable in either input dtype." I haven't thought about if that's a correct definition of this promotion logic, just an example.

We can still reference that this is more like numpy than torch.promote_types and why we're using this logic, but even if we can enumerate the exceptions, the reader shouldn't feel a need to check numpy docs to understand it.

Reply:
Unfortunately we can't guarantee to hold any value when promoting across types, but I'll try to think of a succinct and reasonable description.

Reply:
Definitions in `np.can_cast(input, result, casting='safe')` might provide some ideas, since that seems close to what our promoted type would allow here.

The numpy definition of "safe" is that `‘safe’ means only casts which can preserve values are allowed.`, but that may not mean "exact" values? If we had casting options and associated docs it'd probably be easier to describe. Our default casting rule doesn't match any of the numpy options.

Reply:
That is their definition but it's a lie:

```
a = np.array(9007199254740993)
b = np.array(9007199254740993-1)
print(np.can_cast(np.int64, np.float64, casting='safe'))
print(a == b)
print(a.astype(np.float64) == b.astype(np.float64))
```

will print True (you can "safely" cast int64 to float64), False (the tensors are not equal), and True (the tensors, when cast to float64, are equal). 

You're right they probably intend to say something like: "only casts to a type with a greater than or equal dynamic range are considered 'safe'". But then they would probably allow casting fp16 to bfloat16 for comparison, and that's not what we want, either. (NumPy avoids that particular issue by not supporting bfloat16.)

Reply:
Right, their definition is a bit odd, though possibly clear enough in their case as it contrasts to their other casting rules. I suspect whatever would be the correct definition of 'safe' is what would correctly determine the appropriate common dtype for promotion purposes.

We only have one casting rule, but may want to add more. If we have multiple sets of promotion rules used in different cases, and a promotion rules can be defined as "the 'smallest' dtype to which both inputs can be cast" for some definition of "smallest" and "cast," then we may need to consider this particular promotion rule (and its definition) when we add a "casting" option later.